make disable_with default in submit_tag

DropsOfSerenity opened the pull request:
I heard on the "Bike Shed" podcast that the idea of possibly making disable_with a default for submit_tag to prevent double submission by default, thought I would draft up a pull request for it and take a stab at it.

DropsOfSerenity committed:
Make disable_with default in submit_tag

Prevents double submission by making disable_with the default.

Default disable_with option will only be applied if user has not
specified her/his own disable_with option, whether that is in the
`data-disable-with` string form or the
`:data => { :disable_with => "Saving..." }` hash form. disable_with
will default to the value attribute.

A configuration option was added to opt out of this functionality if
the user so desires.
`config.action_view.automatically_disable_submit_tag = false`

kaspth commented on actionview/lib/action_view/helpers/form_tag_helper.rb:
```diff
@@ -412,36 +412,36 @@ def radio_button_tag(name, value, checked = false, options = {})
         options = options.stringify_keys
 
-        tag :input, { "type" => "submit", "name" => "commit", "value" => value }.update(options)
+        tag :input, { "data" => {"disable_with": "Please wait..."}, "type" => "submit", "name" => "commit", "value" => value }.update(options)
```
Add a space after `{` and before `}`, like the outer hash has.

kaspth replied:
Also we've just gone stringified the keys above. So we need to use a string key here as well.

kaspth commented on actionview/CHANGELOG.md:
```diff
@@ -175,4 +175,9 @@
+*   Make `disable_with` the default behavior for submit tags. Disables the
+    button on submit to prevent double submits.
+
+    *Justin Schiff*
```
New entries go at the top of the file.

kaspth replied:
Can you move your changelog entry to the top of this file? :smile:

DropsOfSerenity commented:
@kaspth updated with your style changes sir :)

rafaelfranca commented on actionview/test/template/form_helper_test.rb:
```diff
@@ -2226,7 +2226,7 @@ def test_submit_with_object_as_new_record_and_locale_strings
 
       expected = whole_form('/posts', 'new_post', 'new_post') do
-        "<input name='commit' type='submit' value='Create Post' />"
+      "<input name='commit' data-disable-with='Please wait...' type='submit' value='Create Post' />"
```
This indentation is wrong now

rafaelfranca commented on actionview/lib/action_view/helpers/form_tag_helper.rb:
```diff
@@ -412,36 +412,36 @@ def radio_button_tag(name, value, checked = false, options = {})
         options = options.stringify_keys
 
-        tag :input, { "type" => "submit", "name" => "commit", "value" => value }.update(options)
+        tag :input, { "data" => { "disable_with" => "Please wait..." }, "type" => "submit", "name" => "commit", "value" => value }.update(options)
```
If someone pass `'data-disable-with'` as options how would it behaves? Also `data: { disable_with: 'foo' }`

rafaelfranca commented:
thank you for the work. It makes sense but I think we may have problems with with the hash update. Could you add tests that it is working with the inputs that I pointed?

DropsOfSerenity commented:
Ok yeah, now that I fixed the tests in `form_tag_helper_test` I got a failing test when any other data attributes were passed in. I fixed by using `deep_merge!` instead of `update` and wrote a test to cover the base case as well.

DropsOfSerenity commented:
ah I see that passing "data-disable-with" to the options has will also result in problems, writing a test and working on a solution.

DropsOfSerenity commented:
Alright I've updated the code and written some tests to ensure that a user specified option is not overridden.

kaspth commented on actionview/lib/action_view/helpers/form_tag_helper.rb:
```diff
@@ -412,36 +412,41 @@ def radio_button_tag(name, value, checked = false, options = {})
+        end
 
-        tag :input, { "type" => "submit", "name" => "commit", "value" => value }.update(options)
+        tag :input, tag_options
```
`submit_tag` has changed enough that I think we need a benchmark to make sure there isn't any serious performance regression.

DropsOfSerenity replied:
Sounds good, I appreciate the feedback, let me take some time tonight solidifying and benchmarking everything.

kaspth commented on actionview/lib/action_view/helpers/form_tag_helper.rb:
```diff
@@ -412,36 +412,41 @@ def radio_button_tag(name, value, checked = false, options = {})
+        tag_options = { "type" => "submit", "name" => "commit", "value" => value }.update(options)
+
+        if not tag_options.include?("data-disable-with") || (tag_options["data"] && tag_options["data"]["disable_with"])
+          tag_options.deep_merge!({ "data" => { "disable_with": "Please wait..." } })
```
This is still creating a symbol in the hash:

```ruby
irb(main):001:0> options = { 'disable_with': 'hello' }
=> {:disable_with=>"hello"}
irb(main):002:0> options[:disable_with]
=> "hello"
```

kaspth commented on actionview/lib/action_view/helpers/form_tag_helper.rb:
```diff
@@ -412,36 +412,41 @@ def radio_button_tag(name, value, checked = false, options = {})
+        options = options.deep_stringify_keys
+        tag_options = { "type" => "submit", "name" => "commit", "value" => value }.update(options)
+
+        if not tag_options.include?("data-disable-with") || (tag_options["data"] && tag_options["data"]["disable_with"])
```
We prefer `unless` over `if not`.

You're checking for "data-disable-with" but your documentation above keeps referring to "data-disable**d**-with". Which is it? :smile:

DropsOfSerenity commented:
@kaspth So I've done some benchmarking and applied a fix.

I found that `deep_stringify_keys` was causing a drop of roughly 30,000 iterations per second, a loss of 37% performance approx.

After changing it to expect the symbol key for the nested hash (which ruby does anyway so it's not necessary to deep stringify) the performance normalized back to 80k ip/s.

```shell
# Rails#master
Calculating -------------------------------------
      submit_tag new     7.988k i/100ms
-------------------------------------------------
      submit_tag new     83.042k (± 2.8%) i/s -    415.376k

# Rails with updated submit_tag code
Calculating -------------------------------------
      submit_tag new     7.711k i/100ms
-------------------------------------------------
      submit_tag new     80.468k (± 2.2%) i/s -    408.683k
```

I think the last thing i'm trying to wrap my head around is that nested hashes by default right now are being stringified only on the base level, and not on the nested level.

e.g.

```ruby
irb(main):004:0> { "data" => { "disable_with": "Please wait..." } }.stringify_keys
=> {"data"=>{:disable_with=>"Please wait..."}}

irb(main):007:0> {}.update({ "data" => { "disable_with": "Please wait..." } }.stringify_keys)
=> {"data"=>{:disable_with=>"Please wait..."}}
```

This would mean this is happening for other data attributes being passed in already as well. Am i missing something here?

bogdan commented:
1. The default string can be longer then original label of submit. It will take extra space when change the label. It can be a problem in tough designs where everything there is no empty space.

2. It needs to be `I18n`-ed by default.

rafaelfranca commented:
> it needs to be I18n-ed by default.

It doesn't. If you need i18n you can just explicitly set `disable_with`.

rafaelfranca commented:
> The default string can be longer then original label of submit. It will take extra space when change the label. It can be a problem in tough designs where there is no empty space.

If that is the case they can just override the default.

rafaelfranca commented:
A problem that I think is: is it possible to remove `disable_with` now that we are setting a default?

DropsOfSerenity commented:
Maybe using "Saving..." instead of "Please wait..." would be a little more device friendly?

sgrif commented:
Based on this implementation, no it doesn't look like there's a way to opt-out. Adding `disable_with: false`, `"data-disable-with": false`, or `data: { "disable-with": false }` should all opt-out of this behavior.

DropsOfSerenity commented:
Alright, I'll add this functionality :)

bogdan commented:
No I18n makes the feature only available for en apps. Others will be forced to specify something everywhere in a project. Also I18n makes it possible to disable the feature and make upgrading easier.

rafaelfranca commented:
We don't have I18n support right now and I want to keep it out. I18n by default will slow down a lot submit_tag for applications that only use english (I bet it is something like 80% of the apps)

bogdan commented:
@rafaelfranca so what should I do after upgrading? Go through my app and specify `disable-with: false` everywhere because now it shows unreadable symbols to all my users. 

Did you include multi-language apps into those 80%? They would also experience problems.

What about people that don't want a change? How would they disable feature globally without modifying entire app?

I think 80% is overestimated: exclude people that don't want this feature and multi-language apps (<- most successful ones by the way).

It looks bad without ability to globally disable. We should have a way of doing it. Also we need somehow tell people in the internationalisation guide:

> Please go to your AppController and extend `submit_tag` to support I18n because we don't support it out of the box.

I hardly image it will work without I18n. It will be top1 change request.

rafaelfranca commented:
> It looks bad without ability to globally disable.

Agree, good point.. I prefer to have it disabled globally without using i18n, but I don't want to introduce other global switch just for that, so I agree that I18n would solve this and also enable other features. :+1:

derekprior commented:
What about using the button text and just disabling the button? If you need something specific, you can override with a string or i18n otherwise the button is just disabled? Not enough user feedback?

DropsOfSerenity commented:
Since our submit button defaults to "Save Changes": `def submit_tag(value = "Save changes", options = {})` I think it would make the most sense to default the text to "Saving..." and allow the user to mess with i18n or define there own disable_with as needed.

bogdan commented:
I like @derekprior's idea. We should not aim to make a good interface - there is no consensus on what label is the best. If we just disable the button, we fix the technical issue that all of agree that needs to be addressed.

derekprior commented:
Don't get me wrong - having a conventional place to go to override the
label without code would be great but if it really is a performance concern
then I think disabling only is a fine compromise.

That's in line with what users see today except that the button is
disabled.

On Friday, August 7, 2015, Bogdan Gusiev <notifications@github.com> wrote:

> I like @derekprior <https://github.com/derekprior>'s idea. We should not
> aim to make a good interface - there is no consensus on what label is the
> best. If we just disable the button, we fix the technical issue that all of
> agree that needs to be addressed.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/rails/rails/pull/21135#issuecomment-128777230>.
>

DropsOfSerenity commented:
Makes sense, in that case we don't want the i18n anymore since it will just be using the value of the submit tag anyway?

DropsOfSerenity commented:
Alright the feature can now be disabled on a per tag basis and uses the `value` passed in as it's default value.

The only thing missing now is if/how we want to be able to disable this feature globally.

DropsOfSerenity commented:
What do you guys think about adding a globally configurable action_view settings like so?

```ruby
# Specify whether submit_tag should automatically disable on click
cattr_accessor :automatically_disable_submit_tag
@@automatically_disable_submit_tag = true
```

So then users that want to disable the feature for whatever reason can set the preference in application.rb
```ruby
config.action_view.automatically_disable_submit_tag = false
```

sgrif commented:
I'm :+1: for adding a config to opt out, if only to aid migration.

On Fri, Aug 7, 2015, 8:26 PM Justin <notifications@github.com> wrote:

> What do you guys think about adding a globally configurable action_view
> settings like so?
>
> # Specify whether submit_tag should automatically disable on click
> cattr_accessor :automatically_disable_submit_tag@@automatically_disable_submit_tag = true
>
> So then users that want to disable the feature for whatever reason can set
> the preference in application.rb
>
> config.action_view.automatically_disable_submit_tag = false
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/rails/rails/pull/21135#issuecomment-128885569>.
>

DropsOfSerenity commented:
Added opt out settings + a test for it.

DropsOfSerenity commented:
Do I need to do anything further for this?

kaspth commented on actionview/lib/action_view/helpers/form_tag_helper.rb:
```diff
@@ -412,36 +412,51 @@ def radio_button_tag(name, value, checked = false, options = {})
       #   disabled version of the submit button when the form is submitted. This feature is
-      #   provided by the unobtrusive JavaScript driver.
+      #   provided by the unobtrusive JavaScript driver. To disable this feature for a single submit tag
+      #   pass <tt>:data => { :disable_with => false }</tt>
```
New hash syntax

kaspth commented on actionview/lib/action_view/helpers/form_tag_helper.rb:
```diff
@@ -412,36 +412,51 @@ def radio_button_tag(name, value, checked = false, options = {})
 
-        tag :input, { "type" => "submit", "name" => "commit", "value" => value }.update(options)
+        if ActionView::Base.automatically_disable_submit_tag
+          disable_with_text = value.clone
```
Writing it this way means we're doing work we don't have to if users have passed disable with themselves.

kaspth commented on actionview/lib/action_view/helpers/form_tag_helper.rb:
```diff
@@ -412,36 +412,51 @@ def radio_button_tag(name, value, checked = false, options = {})
+        if ActionView::Base.automatically_disable_submit_tag
+          disable_with_text = value.clone
+          disable_with_text = tag_options["data"][:disable_with] if tag_options["data"] && tag_options["data"].has_key?(:disable_with)
+          disable_with_text = tag_options["data-disable-with"] if tag_options.has_key?("data-disable-with")
```
How about doing it this way? :smile:

```ruby
disable_with_text = tag_options["data-disable-with"]
disable_with_text ||= tag_options["data"][:disable_with] || tag_options["data"]["disable_with"] if tag_options["data"]
disable_with_text ||= value.clone
```

DropsOfSerenity replied:
We have to allow false to be passed in, ||= will ignore

kaspth replied:
Then the variable name feels a little off :wink: And even so, there should be a simpler way to write the above.

DropsOfSerenity replied:
Ah yeah it's definitely a little unclear, I have tests wrapping it so i'll see what I can do to make it cleared logically and name-wise.

kaspth commented on actionview/lib/action_view/helpers/form_tag_helper.rb:
```diff
@@ -412,36 +412,51 @@ def radio_button_tag(name, value, checked = false, options = {})
+          if disable_with_text
+            tag_options.deep_merge!({ "data" => { "disable_with" => disable_with_text } })
+          else
+            tag_options.extract!("data-disable-with")
```
`extract!` creates and returns a new hash, which we don't care about at all here. It seems we're counting on `delete` being called and should just use that directly (we don't need to care about the `key?` checks, `delete` can handle missing keys).

kaspth commented on actionview/lib/action_view/helpers/form_tag_helper.rb:
```diff
@@ -412,36 +412,51 @@ def radio_button_tag(name, value, checked = false, options = {})
+            tag_options.deep_merge!({ "data" => { "disable_with" => disable_with_text } })
+          else
+            tag_options.extract!("data-disable-with")
+            tag_options["data"].extract!(:disable_with) if tag_options["data"]
```
Use `delete` instead of `extract!` here too.

kaspth commented on actionview/lib/action_view/helpers/form_tag_helper.rb:
```diff
@@ -412,36 +412,51 @@ def radio_button_tag(name, value, checked = false, options = {})
       #   drivers will provide a prompt with the question specified. If the user accepts,
       #   the form is processed normally, otherwise no action is taken.
-      # * <tt>:disable_with</tt> - Value of this parameter will be used as the value for a
+      # * <tt>:disable_with</tt> - Defaults to value attribute. Value of this parameter will be used as the value for a
```
You haven't explained what `:disable_with` does yet, and then you say it defaults to value - how should users know what that means? :smile:

Lets put it last in the description.

DropsOfSerenity replied:
Yeah good point :D

kaspth commented on actionview/lib/action_view/helpers/form_tag_helper.rb:
```diff
@@ -412,36 +412,51 @@ def radio_button_tag(name, value, checked = false, options = {})
+          disable_with_text = tag_options["data-disable-with"] if tag_options.has_key?("data-disable-with")
+
+          if disable_with_text
+            tag_options.deep_merge!({ "data" => { "disable_with" => disable_with_text } })
```
Remove the braces around the `data` hash.

DropsOfSerenity commented:
Ok so in summary of recent changes:

* Used new hash syntax in comment on how to disable.
* Made an initial check to see if the user has disabled the feature on this specific submit tag before doing work.
* Updated the logic using ||= now that the case of the user disabling the feature has been taken care of beforehand.
* Used `Hash#delete` instead of `Hash#extract!`
* Moved explanation of what disable_with defaults to at the end of the doc.
* Removed unnecessary braces.

DropsOfSerenity commented:
Any other ideas guys?

sgrif commented:
Can you add this issue number to the upgrade guide meta issue?

DropsOfSerenity commented:
@sgrif Where can I find that?

sgrif commented:
Just did it. Was on my phone earlier and didn't have time to search.

DropsOfSerenity commented:
Ah thankyou I kind of had no idea what you were talking about since this is my first time contributing to rails :P but thanks :D

claudiofullscreen commented:
It looks like this does not work with the `button_to` helper, where the generated code is something on the line of:

```html
<form class="button_to" method="post" action="…">
  <button type="submit">…</button>
</form>
```

@DropsOfSerenity @sgrif I think we should make that happen… what do you think?

claudiob commented:
It looks like this does not work with the `button_to` helper, where the generated code is something on the line of:

```html
<form class="button_to" method="post" action="…">
  <button type="submit">…</button>
</form>
```

@DropsOfSerenity @sgrif I think we should make that happen… what do you think?
